### PR TITLE
Fix the EMPTY_STRING issue

### PR DESCRIPTION
**Pull request by zhencliu**
ID: 2119339

### CODE REVIEW

**Comment by zhencliu** [comment]
Ran nic_hotplug.one_pci.nic_virtio.default with customsparams='romfile_nic1=EMPTY_STRING'

qemu cmdline:
-device virtio-net-pci,mac=9a:d8:af:77:10:0f,romfile=,id=idfVaEJG,netdev=idqfquQQ,bus=pcie-root-port-3,addr=0x0

hotplug cmdline:
 {'execute': 'device_add', 'arguments': OrderedDict([('id', 'idlZjNuQ'), ('driver', 'virtio-net-pci'), ('netdev', 'idxDMNpL'), ('mac', '9a:d8:af:77:10:10'), ('romfile', ''), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')]), 'id': 'o21bhMiS'}

**Comment by zhencliu** [comment]
hi @PaulYuuu @luckyh , would you pls help to review this PR? Thanks.

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -658,6 +658,10 @@ def add_nic(devices, vlan, model=None, mac=None, device_id=None,
                     if "rombar" in devices.execute_qemu("-device %s,?"
                                                         % model):
                         dev.set_param("rombar", 0)
+
+                # romfile can be '', i.e. to disable rom loading
+                if romfile is not None:
```
Comment (luckyh) [inline virttest/qemu_vm.py:663]: As the target option would be omitted when its value is `None` I think it's safe to remove this check, what do you think?

https://github.com/avocado-framework/avocado-vt/blob/c40074ba6fe4c089f1418e9b44bae413c12bdcdd/virttest/qemu_devices/qdevices.py#L140-L143

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -4001,6 +4007,8 @@ def add_nic(self, **params):
         nic.set_if_none('queues', params.get('queues', '1'))
         if params.get("enable_msix_vectors") == "yes" and int(nic.queues) > 1:
             nic.set_if_none('vectors', 2 * int(nic.queues) + 2)
+        if params.get("romfile") is not None:
```
Comment (luckyh) [inline virttest/qemu_vm.py:4010]: Suggest either `vm_nic_romfile` or `nic_romfile` as the param name for preventing the potential naming collision issue.

File: virttest/qemu_vm.py [hunk 3]
```diff
@@ -4001,6 +4007,8 @@ def add_nic(self, **params):
...
+            nic.set_if_none('romfile', params['romfile'])
```
Comment (luckyh) [inline virttest/qemu_vm.py:4011]: Looks the `.set_if_value_not_none` call would be the one for this case.

File: virttest/qemu_vm.py (hunk 1 above)
Comment (zhencliu) [reply virttest/qemu_vm.py]: Done

File: virttest/qemu_vm.py (hunk 3 above)
Comment (zhencliu) [reply virttest/qemu_vm.py]: Done

File: virttest/qemu_vm.py (hunk 2 above)
Comment (zhencliu) [reply virttest/qemu_vm.py]: In fact, I found that romfile had already been introduced, pls refer to https://github.com/avocado-framework/avocado-vt/blob/c40074ba6fe4c089f1418e9b44bae413c12bdcdd/virttest/utils_net.py#L2676, so if we want to change the param name, I am afraid that could lead to unexpected issues in upstream(checked tp-qemu, no romfile was set in tc)

Comment (zhencliu) [reply virttest/qemu_vm.py]: @luckyh , what's your opinion? Can we replace romfile with nic_romfile?

Comment (PaulYuuu) [reply virttest/qemu_vm.py]: Actually, I don't understand the background, but some extra params can be appended using `nic_extra_params` or `netdev_extra_params`, so we don't need to introduce new parameters for each option.

**Comment by PaulYuuu** [comment]
> Ran nic_hotplug.one_pci.nic_virtio.default with customsparams='romfile_nic1=EMPTY_STRING'
> 
> qemu cmdline: -device virtio-net-pci,mac=9a:d8:af:77:10:0f,romfile=,id=idfVaEJG,netdev=idqfquQQ,bus=pcie-root-port-3,addr=0x0
> 
> hotplug cmdline: {'execute': 'device_add', 'arguments': OrderedDict([('id', 'idlZjNuQ'), ('driver', 'virtio-net-pci'), ('netdev', 'idxDMNpL'), ('mac', '9a:d8:af:77:10:10'), ('romfile', ''), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')]), 'id': 'o21bhMiS'}
> 
> Ran nic_hotplug.multi_pci.nic_virtio.q35 with/without romfile, all passed Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_virtio.q35: PASS

After I checked, even `romfile=` or `romfile=""`, the outputs of `info qtree` are the same.

```
# /usr/libexec/qemu-kvm -netdev tap,id=net1 -device virtio-net-pci,id=nic1,netdev=net1,romfile= -monitor stdio -M q35

(qemu) info qtree
......
......
romfile = ""

# /usr/libexec/qemu-kvm -netdev tap,id=net1 -device virtio-net-pci,id=nic1,netdev=net1,romfile="" -monitor stdio -M q35

(qemu) info qtree
......
......
romfile = ""
```

So if you just want to fix the JSON parsing, I don't think that much modification is needed, only letting json can convert `""` to an empty value is enough. Seems `romfile=""` is more safely in the raw qemu command line.

File: virttest/qemu_vm.py (hunk 2 above)
Comment (zhencliu) [reply virttest/qemu_vm.py]: thanks Yihuang, actually, romfile is not a new param, it is present already, pls refer to my comment above as well as the following:
https://github.com/avocado-framework/avocado-vt/blob/24904ca5335f91a48b31cf6ffce7be287229bc84/virttest/qemu_vm.py#L4135

Background:
For amd sev, we need to disable rom loading by set romfile='', but currently the handling of 'empty string' doesn't work well, so I submitted two commits here to fix it.

**Comment by zhencliu** [comment]
> > Ran nic_hotplug.one_pci.nic_virtio.default with customsparams='romfile_nic1=EMPTY_STRING'
> > qemu cmdline: -device virtio-net-pci,mac=9a:d8:af:77:10:0f,romfile=,id=idfVaEJG,netdev=idqfquQQ,bus=pcie-root-port-3,addr=0x0
> > hotplug cmdline: {'execute': 'device_add', 'arguments': OrderedDict([('id', 'idlZjNuQ'), ('driver', 'virtio-net-pci'), ('netdev', 'idxDMNpL'), ('mac', '9a:d8:af:77:10:10'), ('romfile', ''), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')]), 'id': 'o21bhMiS'}
> > Ran nic_hotplug.multi_pci.nic_virtio.q35 with/without romfile, all passed Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.nic_hotplug.multi_pci.nic_virtio.q35: PASS
> 
> After I checked, even `romfile=` or `romfile=""`, the outputs of `info qtree` are the same.

Correct, it both works fine in the raw format (ie. romfile=xxx)

> 
> ```
> # /usr/libexec/qemu-kvm -netdev tap,id=net1 -device virtio-net-pci,id=nic1,netdev=net1,romfile= -monitor stdio -M q35
> 
> (qemu) info qtree
> ......
> ......
> romfile = ""
> 
> # /usr/libexec/qemu-kvm -netdev tap,id=net1 -device virtio-net-pci,id=nic1,netdev=net1,romfile="" -monitor stdio -M q35
> 
> (qemu) info qtree
> ......
> ......
> romfile = ""
> ```
> 
> So if you just want to fix the JSON parsing, I don't think that much modification is needed, only letting json can convert `""` to an empty value is enough. 

Not exactly.
If we fix it this way, then we might need to handle this for all devices who have options with a valid empty string('') value.
So one commit here is the fix for 'EMPTY_STRING', which can handle the situation that any option of any device has '' as its valid value.

> Seems `romfile=""` is more safely in the raw qemu command line.

I did the checks:
1. -device xxx, romfile=,xxx
2. -device xxx, romfile=
All works well in bash/zsh/csh, so I think it's OK.
And in future, after Nick's json format qemu cmdline patch is merged, for most cases, we will use the json format qemu cmdline

**Comment by PaulYuuu** [comment]
> Not exactly. If we fix it this way, then we might need to handle this for all devices who have options with a valid empty string('') value. So one commit here is the fix for 'EMPTY_STRING', which can handle the situation that any option of any device has '' as its valid value.
> 
> > Seems `romfile=""` is more safely in the raw qemu command line.
> 
> I did the checks:
> 
> 1. -device xxx, romfile=,xxx
> 2. -device xxx, romfile=
>    All works well in bash/zsh/csh, so I think it's OK.
>    And in future, after Nick's json format qemu cmdline patch is merged, for most cases, we will use the json format qemu cmdline

So my thinking is: this issue only happens when using JSON format qemu command line, and not only for nic devices, so in my opinion, it should be considered and handled when we add JSON format support. And for `romfile` option, I think nic_extra_params is enough.

**Comment by zhencliu** [comment]
> > Not exactly. If we fix it this way, then we might need to handle this for all devices who have options with a valid empty string('') value. So one commit here is the fix for 'EMPTY_STRING', which can handle the situation that any option of any device has '' as its valid value.
> > > Seems `romfile=""` is more safely in the raw qemu command line.
> > 
> > 
> > I did the checks:
> > 
> > 1. -device xxx, romfile=,xxx
> > 2. -device xxx, romfile=
> >    All works well in bash/zsh/csh, so I think it's OK.
> >    And in future, after Nick's json format qemu cmdline patch is merged, for most cases, we will use the json format qemu cmdline
> 
> So my thinking is: this issue only happens when using JSON format qemu command line, and not only for nic devices, so in my opinion, it should be considered and handled when we add JSON format support. And for `romfile` option,

Besides qemu-kvm json cmdline, it also affects the qmp cmdline, e.g. hotplug a nic.

> I think nic_extra_params is enough.
It's OK for me to handle it along with nic_extra_params, but refer to the code, currently, romfile is considered as a param of the nic device, pls check my comments https://github.com/avocado-framework/avocado-vt/pull/3491#discussion_r947379347, also please correct me if I didn't understand it correctly. Thanks.

**Comment by zhencliu** [comment]
hi @PaulYuuu @luckyh , talked with Xu this afternoon, I'll divide this PR into two PRs, this one is to fix the empty string issue only, #3494 will handle the romfile issue, thanks for your comments

**Comment by zhencliu** [comment]
@luckyh , please make a double check, thanks.

**Review by luckyh — APPROVED** [review]
LGTM

**Review by PaulYuuu — APPROVED** [review]

**Comment by zhencliu** [comment]
Thank you for your effort, @luckyh , would you please help to merge it? Thanks.

**Pull request merged**